src/commands/registration/pasv.js

```diff
@@ -11,12 +11,17 @@ module.exports = {
     this.connector = new PassiveConnector(this);
     return this.connector.setupServer()
     .then((server) => {
-      let address = this.server.options.pasv_url;
-      // Allow connecting from local
-      if (isLocalIP(this.ip)) {
-        address = this.ip;
-      }
       const {port} = server.address();
+      let pasvAddress = this.server.options.pasv_url;
+      if (typeof pasvAddress === "function") {
+        return Promise.try(() => pasvAddress(this.ip))
+          .then((address) => ({address, port}));
+      }
+      // Allow connecting from local
+      if (isLocalIP(this.ip)) pasvAddress = this.ip;
+      return {address: pasvAddress, port};
+    })
+    .then((address, port) => {
```

Comment:
```suggestion
      return {address: pasvAddress, port};
    })
    .then(({ address, port }) => {
```

Very close! we will need to destructure the object in the next function though - technically we are just returning one parameter. I am actually unsure how to get around this (returning more than one item) in nodejs promise chains.

Reply:
yes, sorry this was a mistake on my side